src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -15,26 +15,122 @@ import scala.annotation.tailrec
 
 trait Ledger {
 
-  def executeBlock(block: Block, validators: Validators): Either[BlockExecutionError, Seq[Receipt]]
+  def executeBlock(block: Block): Either[BlockExecutionError, Seq[Receipt]]
 
-  def prepareBlock(block: Block, validators: Validators): BlockPreparationResult
+  def prepareBlock(block: Block): BlockPreparationResult
 
   def simulateTransaction(stx: SignedTransaction, blockHeader: BlockHeader): TxResult
+
+  def importBlock(block: Block): BlockImportResult
+
+  def resolveBranch(headers: Seq[BlockHeader]): BranchResolutionResult
 }
 
 //FIXME: Make Ledger independent of BlockchainImpl, for which it should become independent of WorldStateProxy type
-class LedgerImpl(vm: VM, blockchain: BlockchainImpl, blockchainConfig: BlockchainConfig) extends Ledger with Logger {
+class LedgerImpl(vm: VM, blockchain: BlockchainImpl, blockchainConfig: BlockchainConfig, validators: Validators) extends Ledger with Logger {
 
   val blockRewardCalculator = new BlockRewardCalculator(blockchainConfig.monetaryPolicyConfig)
 
-  def executeBlock(
-    block: Block,
-    validators: Validators): Either[BlockExecutionError, Seq[Receipt]] = {
+  def importBlock(block: Block): BlockImportResult = {
+    // TODO: ignore blocks that are too far behind the head
+    if (blockchain.getBlockByHash(block.header.hash).isDefined)
+      ImportFailedBlockExists
+
+    else
+      blockchain.getBlockByHash(block.header.parentHash) match {
+        case None =>
+          ImportFailedMissingParent
```

Comment:
Should we accept orphaned blocks and keep them in a queue of some sort for later execution? It looks to me like other clients do something of such sort (that's as specific as I can get ATM).